### ISSUE TITLE
Lagging and Crashing after working for long periods of time

### ISSUE THREAD

**Issue by tensor-programming** [Dart-Code/Dart-Code#1721]
I am on the Flutter dev 1.6.0 version with windows 10 on VSCode Stable branch.  

After updating to this version of flutter today, I was finding that the Dart Code extension was lagging and "crashing" after working for about 30-40 minutes worth of time.  The extension wasn't crashing the editor but it would basically stop working to a point where I would have to go into my task manager and kill the dart instances.  speaking of which, there seemed to be 2-3 open at once at all times while the analyzer was running and often times they would be taking up 100% of my memory (64 gigs on this machine) when these crashes happened.  It was really annoying and isn't something that I've ever specifically encountered.     

After killing the Dart instances, I would have to restart the analyzer (and sometimes the editor) which would buy me another 30-40 minutes worth of time. 

This all happened during one of my live streams though you can't really see whats happening since I kept changing windows when I had to restart the editor.  In the entire 2 hour and 30 minute session, I had to restart the editor 3-4 times. 

Edit: Running the code in the Emulator and using DevTools didn't seem to change anything or tax the machine anymore then the normal analyzer.

**Comment by DanTup** [comment]
I scanned through your stream and noticed it said "Analyzing..." the whole time in the status bar. I wonder if you're hitting https://github.com/Dart-Code/Dart-Code/issues/1713. Do you have the `dart.analyzerObservatoryPort` setting set? If so, please unset it and see if the problem goes away.

If not, please [capture a log](https://dartcode.org/docs/logging/#capture-logs-command) ticking the Analysis Server to see if there's anything obvious in there.

That said - this doesn't explain the processes using 64GB RAM. Are you able to see exactly which Dart process is using it (what the command line was it was started with)? We may need to run it with Observatory (which is the thing the setting above will disable) and use the memory profiling to figure out what's using all the RAM.

**Comment by tensor-programming** [comment]
I pushed my flutter version back to stable due to another issue that I encountered (issue with android alarm manager).  If the error doesn't persists on the stable branch, I'll go back to the dev branch (`1.6.0`) and see if I can capture a log and mark the offending process for reference.   

Currently, I don't have a `dart.analyzerObservatoryPort` set and there is nothing obvious that would account for this issue.

**Comment by DanTup** [comment]
Thanks - I'll keep my eyes out for similar on `master`.

If you're not show when it'll come up, you could enable the [analysis server log](https://dartcode.org/docs/logging/#analyzer) and leave it enabled until you see the issue again (it'll overwrite the log file on each new session, so won't be kept indefinitely), then grab a copy of the log.

**Comment by tensor-programming** [comment]
Alright, I will keep that in mind and hopefully get you a decent log if there is still an error present.

**Comment by DanTup** [comment]
I think https://github.com/flutter/flutter/issues/33280 is the same as this, and I believe I saw it last night (and got a log, but it's very large). I'll do some more digging...

**Comment by tensor-programming** [comment]
I appreciate that.  I've been doing development fairly sparingly and I haven't had this issue again but I am doing a live stream tomorrow so I should have some idea of whats going on at that point.  I feel like I need to have an extended period of time using the analyzer before I can really see the problem again.

**Comment by tensor-programming** [comment]
After the live stream, I did indeed run into this issue.  It seemed like the Analyzer was slowly leaking memory or something because I had to restart it about 40-50 minutes in.  I does seem to be related to [#flutter/flutter#33280](https://github.com/flutter/flutter/issues/33280).  I wasn't able to give you a log because the error didn't quite happen again by the end of the stream.  Ill see if I can get one in the coming week if you need one.

**Comment by DanTup** [comment]
I saw this again last night after a few hours of coding on a Flutter project ((tiler)[https://github.com/DanTup/tiler]) for several hours. The editor started crawling and I had analysis errors from while I was typing persist for quite a while after.

Dev console appeared to report lots of leaking handlers:

![code perf issues](https://user-images.githubusercontent.com/1078012/58405103-936a2900-805e-11e9-88fc-d155cfc25170.png)

`code --status` showed the extension host eating CPU:

```

Version:          Code 1.34.0 (a622c65b2c713c890fcf4fbf07cf34049d5fe758, 2019-05-15T21:59:37.030Z)
OS Version:       Windows_NT x64 10.0.17763
CPUs:             Intel(R) Core(TM) i5-4430 CPU @ 3.00GHz (4 x 2999)
Memory (System):  15.93GB (6.03GB free)
VM:               0%
Screen Reader:    no
Process Argv:     
GPU Status:       2d_canvas:                    enabled
                  checker_imaging:              disabled_off
                  flash_3d:                     enabled
                  flash_stage3d:                enabled
                  flash_stage3d_baseline:       enabled
                  gpu_compositing:              enabled
                  multiple_raster_threads:      enabled_on
                  native_gpu_memory_buffers:    disabled_software
                  rasterization:                enabled
                  surface_synchronization:      enabled_on
                  video_decode:                 enabled
                  webgl:                        enabled
                  webgl2:                       enabled

CPU %	Mem MB	   PID	Process
    0	    82	 10684	code main
    1	   147	  4552	   gpu-process
    0	    88	  6404	   window (mapobjectitem.cpp - tiled - Visual Studio Code)
    0	    12	   692	     watcherService 
    0	     5	  7700	       console-window-host (Windows internal process)
    0	    20	  9188	     searchService
    0	    28	 10028	     extensionHost
    0	   263	  7544	   window (Running Extensions - tiler - Visual Studio Code)
    0	     5	  1968	     winpty-process
    0	     7	  6828	       console-window-host (Windows internal process)
    0	    56	 10872	       C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe
   24	  2392	  3180	     extensionHost
    0	     3	  5812	       C:\Windows\system32\cmd.exe /d /s /c ""M:\Dev\Google\flutter\bin\cache\dart-sdk\bin\dart.exe" "M:\Dev\Google\flutter\bin\cache\dart-sdk\bin\snapshots\analysis_server.dart.snapshot" "--client-id=Dart-Code.dart-code" "--client-version=3.0.2""
    0	     5	 10460	         console-window-host (Windows internal process)
    0	   357	 10696	         "M:\Dev\Google\flutter\bin\cache\dart-sdk\bin\dart.exe"  "M:\Dev\Google\flutter\bin\cache\dart-sdk\bin\snapshots\analysis_server.dart.snapshot" "--client-id=Dart-Code.dart-code" "--client-version=3.0.2"
    0	     3	  8352	       C:\Windows\system32\cmd.exe /d /s /c ""M:\Dev\Google\flutter\bin\flutter.bat" "run" "--machine" "-t" "lib\main.dart" "-d" "emulator-5554" "--track-widget-creation" "--start-paused""
    0	     5	  1208	         console-window-host (Windows internal process)
    0	   136	  5128	         "M:\Dev\Google\flutter\bin\cache\dart-sdk\bin\dart.exe"   "M:\Dev\Google\flutter\bin\cache\flutter_tools.snapshot" "run" "--machine" "-t" "lib\main.dart" "-d" "emulator-5554" "--track-widget-creation" "--start-paused" 
    0	   254	  3288	           M:\Dev\Google\flutter\bin\cache\dart-sdk\bin\dart.EXE M:\Dev\Google\flutter\bin\cache\artifacts\engine\windows-x64\frontend_server.dart.snapshot --sdk-root M:\Dev\Google\flutter\bin\cache\artifacts\engine\common\flutter_patched_sdk/ --incremental --strong --target=flutter --output-dill build\app.dill.track.dill --packages M:\Dev\tiler\example\.packages --track-widget-creation
    0	     6	  3632	           N:\Apps\Android\platform-tools\adb.exe -s emulator-5554 shell -x logcat -v time
    0	    21	  9164	       "C:\Users\danny\AppData\Local\Programs\Microsoft VS Code\Code.exe" "c:\Users\danny\AppData\Local\Programs\Microsoft VS Code\resources\app\extensions\json-language-features\server\dist\jsonServerMain" --node-ipc --clientProcessId=3180
    0	     3	 10396	       C:\Windows\system32\cmd.exe /d /s /c ""M:\Dev\Google\flutter\bin\flutter.bat" "daemon""
    0	     5	  1836	         console-window-host (Windows internal process)
    0	    32	 10352	         "M:\Dev\Google\flutter\bin\cache\dart-sdk\bin\dart.exe"   "M:\Dev\Google\flutter\bin\cache\flutter_tools.snapshot" "daemon" 
    0	     5	 10216	           N:\Apps\Android\emulator\emulator.exe -avd flutter_emulator
    1	  1456	  4436	             N:\Apps\Android\emulator\qemu\windows-x86_64\qemu-system-i386.exe -avd flutter_emulator
    0	     3	  2980	               /C "N:\Apps\Android\emulator\emulator64-crash-service.exe -pipe \\.\pipe\AndroidEmulatorCrashService.4436 -ppid 4436 -data-dir "X:\Danny Temp\\AndroidEmulator\8dfe68af-7feb-11e9-8330-d05099106836" >nul 2>&1"
    0	     8	 10056	                 N:\Apps\Android\emulator\emulator64-crash-service.exe  -pipe \\.\pipe\AndroidEmulatorCrashService.4436 -ppid 4436 -data-dir "X:\Danny Temp\\AndroidEmulator\8dfe68af-7feb-11e9-8330-d05099106836" 
    0	     0	  9796	               
    0	    10	  5600	     electron-crash-reporter
    0	    15	  7816	     watcherService 
    0	     5	  7600	       console-window-host (Windows internal process)
    0	    67	 10532	   shared-process

Workspace Stats: 
|  Window (Running Extensions - tiler - Visual Studio Code)
|  Window (mapobjectitem.cpp - tiled - Visual Studio Code)
|    Folder (tiled): 1495 files
|      File types: png(382) h(319) cpp(310) tmx(54) qbs(34) java(27) pro(27)
|                  ts(26) rst(23) po(20)
|      Conf files: makefile(1)
|    Folder (tiler): 308 files
|      File types: png(101) json(45) dart(14) txt(14) xml(11) bin(10) dex(9)
|                  flat(7) properties(7) lock(7)
|      Conf files: launch.json(1)

```

Before I was able to capture a profile of the extension host, it crashed and restarted.

I'll try doing some profiling today to see if I can identify any issues without leaving it to get to the point of crashing!

**Comment by DanTup** [comment]
I've been doing some testing today, and although I can't reproduce memory running away, the performance of completions in large projects (esp. those refrencing Flutter) is unacceptably bad. In the next release (expected to ship at the end of the week) I'm [disabling completion for unimported files](https://github.com/Dart-Code/Dart-Code/issues/1751) by default (you can still opt-in if you really want) while I work on improving the performance and reducing memory allocations.

**Comment by DanTup** [comment]
I made a new beta at https://github.com/Dart-Code/Dart-Code/releases/tag/v3.1.0-beta.2 that includes:

1. Turning off the auto-import completions by default (you can still opt back in)
2. Handles when VS Code cancels a request for completions (eg. if the user clicks away before we finish building the list) which should save some processing from all the JSON serialisation/deserialisation for a request that's stale

I think (2) may help, but I don't think it solves the whole issue. While doing some investigation today, I found that building the completion list is actually not that slow, the biggest contributor to both CPU and memory usage seems to be VS Code getting the results from the extension host back to the UI. I've opened https://github.com/microsoft/vscode/issues/74418 to see if there's any way that can be improved.

**Comment by tensor-programming** [comment]
Ill see if the beta build improves things on my next flutter stream.

**Comment by DanTup** [comment]
Good news/bad news...

Good news: Got to the bottom of this - identified a memory leak in VS Code: https://github.com/microsoft/vscode/issues/74446
Bad news: neither disabling auto-import completions nor the cancellation fix in the beta will prevent it (though both will reduce its impact)

I expect Code will get a fix in its next release (which I think is 1-2 weeks away), and hopefully it'll be in the Insiders builds from tomorrow (assuming they get a fix today). As such, I'm reverting my change to disable this by default in vNext.

Recommendations till it ship are still to disable the auto-import completions (or use the beta that has them disabled - though note the next beta will re-enable them by default), or use Insiders (once it has the fix).

Thanks for your help digging into this! :-)

**Comment by DanTup** [comment]
I've tested the VS Code changes in the Insiders build and the memory leak in completions has gone, so I'm considering this fixed (though unless you want to use Insiders, you won't see the fix until the next VS Code release (which [this issue](https://github.com/microsoft/vscode/issues/74412) says is targeting June 6).

Should you see this again using v1.35 in VS Code, then of course please let me know!

**Comment by tensor-programming** [comment]
Very good, @DanTup, still experiencing it with the current build.  Looking forward to the release in June.

**Comment by DanTup** [comment]
VS Code released yesterday, so if you update to v1.35 hopefully you won't see this again (but if you do, please let me know).
